### PR TITLE
feat(cta): add rh-cta

### LINKED ISSUE #333: [feat] <rh-cta>
### Description

`<pfe-cta>` should be migrated to RHDS. when `<rh-cta>` is ready, `<pfe-cta>` will no longer be updated.

### Acceptance Criteria

- [ ] Design done
- [ ] Development done

### Image

_No response_

### Link to design doc

_No response_

### Other resources

See #238, design should cover those edge cases

### PR DESCRIPTION
**Pull request by bennypowers**
## What I did

1. implement rh-cta
2. add color context to `lib` (will eventually be removed from pfe, so import from here from now on)
3. add `DirController` - useful when you need to imperatively style things based on rtl

### Testing Instructions
- Examine each interaction state (default, hover, focus, active) for each CTA in https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/
- `npm ci && npm run test:e2e`. Inspect the test files generated in `element/rh-cta/test/`
- @heyMP please check the integration with footer
- @zeroedin please check the integration with secondary nav. We might want to migrate secondary nav to use the local context decorators, but if this PR works, we can do that later.
- we're missing the video icon, so i didn't implement it. see discussion

## Notes to reviewers
- Please carefully review the token names used in the variants section of rh-cta.css - do all the token names used represent their semantic use? i.e. did I use border color tokens for instances of border?
- The brick-on-dark styles were not specified, so I improvised
- there is no `on saturated` as it wasn't specified
### Old Notes
This depends a lot on design-tokens, and there are lots of unanswered questions there at the time of this writing.
The biggest single feature of CTA is theming and context (which also includes `broadcast`), and it's not clear what the future holds for those concepts in both pfe and rhds. Please see the relevant discussion:
- https://github.com/RedHat-UX/red-hat-design-tokens/pull/2
- https://github.com/orgs/RedHat-UX/discussions/348
- https://github.com/orgs/RedHat-UX/discussions/306
Ensure this fixes https://github.com/RedHat-UX/red-hat-design-system/issues/238

Closes #333

### CODE REVIEW

**Comment by bennypowers** [comment]
@marionnegp would you please take a preliminary look at the arrow for the default cta at https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/

The change I made was to animate the `translate` value instead of the `padding` value, which is much more performant, but will perhaps change some of the pixel values, especially in relationship with sibling elements or containing elements

**Comment by marionnegp** [comment]
I agree that it makes more sense to use `translate`! The spacing between the arrow and the text is supposed to be 8px. It looks like if you use `padding-inline: 5px;` for the arrow, it'll have the right spacing, but it's a tiny tweak.

**Comment by bennypowers** [comment]
@RedHat-UX/designers where does the video play icon on ux.redhat.com/components/icon come from?

**Comment by bennypowers** [comment]
If reviewers with to merge, please squash

**Comment by heyMP** [comment]
@bennypowers thanks for all of your work on this. Looks like there is a difference in size and dark color context when rh-cta is in rh-footer.  That we'll need to investigate:

Screen Shot 2022-08-16 at 8 22 09 PM

**Comment by bennypowers** [comment]
@heyMP checking [the deploy preview](https://deploy-preview-349--red-hat-design-system.netlify.app/components/footer/demo/) as well as the dev server, the footer demo using rh-cta appears correct to me. Can you confirm? How can I reproduce your results from above?

**Comment by bennypowers** [comment]
I removed the built-in analytics events from cta, as described in https://github.com/RedHat-UX/red-hat-design-system/issues/333#issuecomment-1209653112

The alternative is IMO much better: analytics users implement their own analytics. This way, cta users without analytics requirements do not need to load and run unnecessary javascript, and analytics users can easily capture cta click events.

I've added a demo page for analytics events, where you can peep the technique, including a javascript snippet which analytics users can copy into their analytics scripts:

https://deploy-preview-349--red-hat-design-system.netlify.app/components/cta/demo/analytics/#primary

paging @kylebuch8, @markcaron

File: elements/rh-cta/demo/analytics.html [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+<script src="https://unpkg.com/prismjs@1.28.0/plugins/copy-to-clipboard/prism-copy-to-clipboard.min.js"></script>
+
+<link rel="stylesheet" href="../demo.css" />
+<script type="module" src="../rh-cta.js"></script>
+<script type="module" src="../analytics.js"></script>
+<script type="module" src="https://unpkg.com/@power-elements/json-viewer?module"></script>
+<script type="module" src="https://unpkg.com/zero-md/dist/zero-md.min.js"></script>
+
+<p>In this demo, analytics events involving <code>&lt;rh-cta></code> elements are parsed by a 
+  document-level analytics event listener. Unlike <code>&lt;hfe-cta></code>, which implemented
```
Comment (zeroedin) [inline elements/rh-cta/demo/analytics.html:16]: Small typo
```Suggestion
  document-level analytics event listener. Unlike <code>&lt;pfe-cta></code>, which implemented
```

File: elements/rh-cta/rh-cta.css [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  padding: 0px;
+  text-align: left;
+}
+
+#container {
+  display: block;
+  white-space: nowrap;
+  min-width: 100%;
+  color: var(--rh-cta-color);
+  font-family: var(--rh-font-family-heading, "RedHatDisplay", "Overpass", Helvetica, Arial, sans-serif);
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply elements/rh-cta/rh-cta.css:40]: see https://github.com/RedHat-UX/red-hat-design-tokens/pull/18
TL;DR: it might end up being `RedHatDisplay, "Red Hat Display"`, etc.
I'd like to write some automation that would update our components' css whenever we update the tokens, but we could do this manually as well

**Comment by zeroedin** [comment]
I adjusted these containers and was able to get the arrow to move to its own line.

image

**Comment by bennypowers** [comment]
> I adjusted these containers and was able to get the arrow to move to its own line.
> 
> image

chrome macos?

**Comment by zeroedin** [comment]
> > I adjusted these containers and was able to get the arrow to move to its own line.
> > image
> 
> chrome macos?

Yes sorry, that info would be helpful.

**Comment by bennypowers** [comment]
alright @zeroedin, enjoy the fix for that one :wink:

File: elements/rh-footer/rh-global-footer.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import type { ColorPalette } from '@patternfly/pfe-core';
 import { LitElement, html } from 'lit';
```
Comment (zeroedin) [inline elements/rh-footer/rh-global-footer.ts:3]: Missed import link for ColorPalette RHDS lib instead of pfe-core

```
import type { ColorPalette } from '../../lib/context/color.js';
```

**Comment by bennypowers** [comment]
:warning: This branch has been **REBASED**

File: elements/rh-footer/rh-global-footer.ts (hunk 3 above)
Comment (bennypowers) [reply elements/rh-footer/rh-global-footer.ts:3]: done

File: elements/rh-footer/demo/rh-footer.html [hunk 4]
```diff
@@ -332,7 +332,7 @@ <h3 slot="header">About Red Hat</h3>
         </rh-footer-block>
         <rh-footer-block slot="main-secondary">
           <h3 slot="header">Subscribe to our free newsletter, Red Hat Shares</h3>
-          <pfe-cta><a href="#blocks">Sign up now</a></pfe-cta>
+          <rh-cta><a href="#blocks">Sign up now</a></rh-cta>
```
Comment (bennypowers) [inline elements/rh-footer/demo/rh-footer.html:335]: Reserving this comment to block auto-merge pending @heyMP's [review](https://github.com/RedHat-UX/red-hat-design-system/pull/349#issuecomment-1217305315).

File: elements/rh-cta/demo/analytics.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+document.addEventListener('click', function(event) {
+  const cta = deepClosest(event, 'rh-cta');
+  if (cta) {
+    const { href, text, title } = cta.cta;
+    const color = cta.colorPalette;
+    const type = cta.variant ?? 'default';
+    const data = { href, text, title, color, type };
+    // eslint-disable-next-line no-console
+    console.log('CTA ANALYTICS EVENT', data);
```
Comment (bennypowers) [inline elements/rh-cta/demo/analytics.js:31]: Reserving this merge-blocking comment for @rmessickSDI to review the approach taken. Resolve this (and other) comments to unblock auto-merge.

**Comment by zeroedin** [comment]
> alright @zeroedin, enjoy the fix for that one 😉

`&nbsp;` nice... 🙈

**Comment by heyMP** [comment]
> @heyMP checking [the deploy preview](https://deploy-preview-349--red-hat-design-system.netlify.app/components/footer/demo/) as well as the dev server, the footer demo using rh-cta appears correct to me. Can you confirm? How can I reproduce your results from above?

Yeah, I don't know why I was seeing something different.  pfe-cta looks great with footer!

File: elements/rh-cta/rh-cta.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * Call to action stands out from regular hypertext links, and is used for linking users to webpages.
+ *
+ * @summary Directs a user to other pages or displays extra content
+ *
+ * @slot - We expect an anchor tag, `<a>` with an `href`, to be the first child inside `rh-cta` element. Less preferred but allowed for specific use-cases include: `<button>` (note however that the `button` tag is not supported for the default CTA styles).
+ *
+ * @fires {CtaSelectEvent} select - This event is fired when a link is clicked and serves as a way to capture click events if necessary.
+ *
```
Comment (heyMP) [inline elements/rh-cta/rh-cta.ts:46]:
```suggestion
```

We should remove this event as discussed here: https://github.com/RedHat-UX/red-hat-design-system/issues/333#issuecomment-1209653112

File: elements/rh-footer/demo/rh-footer.html (hunk 4 above)
Comment (heyMP) [reply elements/rh-footer/demo/rh-footer.html:335]: Works perfectly 👍

File: lib/context/color.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+);
+
+export interface ColorContextOptions {
+  prefix?: string;
+  attribute?: string;
+}
+
+// TODO: CSS
+// 1. move sass that maps from palette to theme from _colors.scss:198+ to color-context.scss (and rename them)
+// 2. except don't because hard, wait for design tokens instead
```
Comment (heyMP) [inline lib/context/color.ts:46]: 😂

File: elements/rh-cta/rh-cta.ts (hunk 6 above)
Comment (bennypowers) [reply elements/rh-cta/rh-cta.ts]: removed

File: elements/rh-cta/demo/analytics.js (hunk 5 above)
Comment (bennypowers) [reply elements/rh-cta/demo/analytics.js:31]: We still want analytics review and feedback but we're moving forward with the PR as is. Please open new issues for any changes needed.

**Review by marionnegp — APPROVED** [review]

**Review by marionnegp — COMMENTED** [review]
- [ ] All borders should be 1px
- [ ] When an inner border appears (for focus and active states in primary, secondary, and brick), there should be 1px of space between it in and the outer border, like this:
brick-cta

- [ ] Brick font-weight looks heavier than 400, but maybe that's a discrepancy between how it displays in demo and on prod.
- [ ] Video icon isn't showing up on demo.
